Add generate-test-data script for mockDiscovery_randomized test

wafschneider opened the pull request:
@ihardy see if this works for your purposes. The `generate-test-data.pl` script takes about 6 minutes to generate a randomized file of instance HRIDs off our test system. To use:

```
Usage:

perl generate-test-data.pl [--help] --config <config file> [output directory]

The script will generate test data in the output directory. Instance
HRIDs in the file "instances.csv" and Okapi connection parameters in
"config_random.csv". Files with those names in the output directory
will be overwritten.

Note that the HRID parsing routine is not perfect and will not
necessarily pick up all HRIDs, depending on how the server chunks the
HTTP response.

Options:

--help | -h : Print help message.

--config | -c : Path to configuration file (required).

Configuration file:

The config file is a simple JSON file using the following format:

{
  "okapi": "[URL of Okapi server]",
  "tenant": "[Okapi tenant ID]",
  "user": "[FOLIO username]",
  "password": "[FOLIO password]",
  "instanceQuery": "[CQL query to retrieve items, optional]",
  "instanceCount": 100000
}

The okapi, tenant, user, and password properties are required. The
other properties are optional. Any CQL queries are ANDed to the
default queries (see below). The defaults for other optional
properties are as shown above.

Default instanceQuery: (cql.allRecords=1)
```

This is the config file I'm using for our test system:

```json
{
  "okapi": "https://lehigh-test-okapi.dev-us-east-2.indexdata.com",
  "tenant": "lu",
  "user": "lehigh_admin",
  "password": "*****"
}
```

(password is in AWS Secrets or in `id-folio-infrastructure`)

The script will generate `instances.csv` and `config_random.csv`. I also did some refactoring of the test to make the number of threads in the main loop configurable, and generate a sample for a resultset page (default 15 instances per page) -- so do take a look at the test plan and make sure it makes sense.

wafschneider committed:
Add generate-test-data script for mockDiscovery_randomized test.

wafschneider committed:
Fix default user variable settings in mockDiscovery_randomized test

ihardy commented:
Thanks @wafschneider the setup script looks Great. I'm not 100% sure I understand how the test works but I'ts highly possible Its flying over my head. So what I gather is that each user defined in global_vusers will read one line of the instances.csv. So w/the default config supplied, we'll get an infinite loop which asks for the same instance over and over until the test user hits stop. If we scale up the number of users, we'll get a number of infinite loops requesting a different instance each equal to the number of users. 

Let me know if that's what you expect the test to do, or if we should adjust it a bit (or if I've completely misinterpreted it!).

Would it make sense to set the loop count to 1 instead of infinite. Then you'd show how many instances you can render at once. So a test with users = 30 would be the real world equivalent of having 2 users load a page of 15 results?

Or would it makes sense to have each thread resolve 15 different instances if that's possible?

wafschneider committed:
Fix CSV data set element in mock-discovery. Clean up loops and other minor fixes.

wafschneider commented:
I made the following changes:
1. Put the CSV data set element in the "resultset" loop, so that each run of that loop gets a new instance HRID
2. Added "Response code" assertions to all requests as part of the main discovery thread group
3. Removed redundant loop controller within main discovery thread group
4. Removed disabled elements and debugging elements

In my testing, it now appears that each run of the "Resultset" loop picks up a new instance HRID. The HRID doesn't appear to be repeated, even with multiple threads.

@ihardy check it out and see if you see the same behavior. Thanks, and sorry for the hassle!

wafschneider committed:
Add documentation for mock-discovery.

ihardy commented:
Tested, and after having Wayne patiently explain I have slowly come to an understanding. Looks great.
